File: golang/lexer/lexer_test.go
package lexer

import (
	"testing"
	"writing_a_interpreter_in_go/token"
)

func TestNextToken(t *testing.T) {
	input := `=+(){},;`

	tests := []struct {
		expectedType    token.TokenType
		expectedLiteral string
	}{
    {token.ASSIGN, "="},
    {token.PLUS, "+"},
    {token.LPAREN, "("},
    {token.RPAREN, ")"},
    {token.LBRACE, "{"},
    {token.RBRACE, "}"},
    {token.COMMA, ","},
    {token.SEMICOLON, ";"},
  }

  l := New(input)

  for i, tt := range tests {
    tok := l.NextToken()

    if tok.Type != tt.expectedType {
      t.Fatalf("test[%d] - tokentype wrong, expeceted=%q, got=%q", i, tt.expectedType, tok.Type)
    }
    if tok.Literal != tt.expectedLiteral {
      t.Fatalf("test[%d] - literal wrong, expeceted=%q, got=%q", i, tt.expectedLiteral, tok.Literal)
    }
  }
}

File: typescript/lexer.test.ts
import { expect, test } from "bun:test";
import { Token, TokenType } from "./token";
import { Lexer } from "./lexer";

type expectedType = {
  expectedType: TokenType;
  expectedLiteral: string;
};

test("Test Next Token", () => {
  const input = `=+(){},;`;

  const tests = [
    {
      expectedType: TokenType.ASSIGN,
      expectedLiteral: "=",
    },
    {
      expectedType: TokenType.PLUS,
      expectedLiteral: "+",
    },
    {
      expectedType: TokenType.LPAREN,
      expectedLiteral: "(",
    },
    {
      expectedType: TokenType.RPAREN,
      expectedLiteral: ")",
    },
    {
      expectedType: TokenType.LBRACE,
      expectedLiteral: "{",
    },
    {
      expectedType: TokenType.RBRACE,
      expectedLiteral: "}",
    },
    {
      expectedType: TokenType.COMMA,
      expectedLiteral: ",",
    },
    {
      expectedType: TokenType.SEMICOLON,
      expectedLiteral: ";",
    },
  ] as expectedType[];

  const lexer = new Lexer(input);

  tests.forEach(({ expectedType, expectedLiteral }) => {
    const token = lexer.nextToken();
    expect(token.tokenType).toBe(expectedType);
    expect(token.literal).toBe(expectedLiteral);
  });
});
